plugin/https/proxy.go

```diff
@@ -0,0 +1,145 @@
+	Query(ctx context.Context, dnsreq []byte) (result *dns.Msg, err error)
+}
+
+// newDoHDNSClient creates a new instance of dohDNSClient service.
+// url must be a full URL to send DoH requests to like "https://example.com/dns-query"
+func newDoHDNSClient(client httpRequestDoer, url string) *dohDNSClient {
+	return &dohDNSClient{client, url}
+}
+
+type httpRequestDoer interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+// dohDNSClient is a DNS client that proxies requests to the upstream server using DoH protocol.
+type dohDNSClient struct {
+	client httpRequestDoer
+	url    string
+}
+
+func (c *dohDNSClient) Query(ctx context.Context, dnsreq []byte) (result *dns.Msg, err error) {
+	var req *http.Request
+	if req, err = http.NewRequestWithContext(ctx, "POST", c.url, bytes.NewReader(dnsreq)); err != nil {
+		return
+	}
+	req.Header["Accept"] = dnsMessageMimeTypeHeader
+	req.Header["Content-Type"] = dnsMessageMimeTypeHeader
+
+	var resp *http.Response
+	if resp, err = c.client.Do(req); err != nil {
+		return
+	}
+	defer resp.Body.Close()
+
+	// RFC8484 Section 4.2.1:
+	// A successful HTTP response with a 2xx status code is used for any valid DNS response,
+	// regardless of the DNS response code.
+	// HTTP responses with non-successful HTTP status codes do not contain
+	// replies to the original DNS question in the HTTP request.
+	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
+		return nil, errResponseStatus
+	}
+
+	// limit the number of bytes read to avoid potential DoS attacks.
+	// it would be better to add (*dns.Msg) Unpack(io.Reader) method to avoid byte slice allocation
+	var body []byte
+	if body, err = io.ReadAll(io.LimitReader(resp.Body, maxDNSMessageSize+1)); err != nil {
+		return
+	}
+	if len(body) > maxDNSMessageSize {
+		return nil, errResponseTooLarge
+	}
+	result = new(dns.Msg)
+	err = result.Unpack(body)
+	return
+}
+
+func newLoadBalanceDNSClient(clients []dnsClient, opts ...lbDNSClientOption) *lbDNSClient {
+	c := &lbDNSClient{
+		p:        newRandomPolicy(),
+		maxFails: len(clients),
+		timeout:  defaultRequestTimeout,
+		clients:  clients,
+	}
+	// option pattern
+	for _, o := range opts {
+		o(c)
+	}
+	if len(clients) < c.maxFails {
+		c.maxFails = len(clients)
+	}
+	return c
+}
+
+// lbDNSClient is a DNS client that load balances DNS requests between the list of DNS clients.
+type lbDNSClient struct {
+	p        policy
+	timeout  time.Duration
+	maxFails int
+	clients  []dnsClient
+}
+
+type lbDNSClientOption func(c *lbDNSClient)
+
+func withLbPolicy(p policy) lbDNSClientOption {
+	return func(c *lbDNSClient) {
+		c.p = p
+	}
+}
+
+func withLbRequestTimeout(timeout time.Duration) lbDNSClientOption {
+	return func(c *lbDNSClient) {
+		c.timeout = timeout
+	}
+}
+
+func withLbMaxFails(maxFails int) lbDNSClientOption {
+	return func(c *lbDNSClient) {
+		c.maxFails = maxFails
+	}
+}
+
+func (c *lbDNSClient) Query(ctx context.Context, dnsreq []byte) (result *dns.Msg, err error) {
```

Comment:
just select the target first and then use that in Query, no need to make this into something special